File: leetcode/84_largest_rectangle_in_histogram.go
package main

type Rect struct {
	X int // x coordinate of the start position
	H int // height of the rectangle
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func largestRectangleArea(heights []int) int {
    n, maxArea := len(heights), 0
    stack := make([]Rect, 0)

	for i := 0; i <= n; i++ {
		startPos := i

        for len(stack) != 0 && (i == n || stack[len(stack)-1].H > heights[i]) {
            prevRect := stack[len(stack)-1]
			maxArea = max(maxArea, (i-prevRect.X)*prevRect.H)
			stack = stack[:len(stack)-1]  // stack.Pop
            startPos = prevRect.X
		}
        
        if i != n {
            stack = append(stack, Rect{startPos, heights[i]})  // stack.Push
        }
	}

	return maxArea
}

File: leetcode/ruby/84_largest_rectangle_in_histogram.rb
def largest_rectangle_area(heights)
  stack = []
  best = 0
  n = heights.length

  0.upto(n).each do |idx|
    start_pos = idx
    while !stack.empty? && (idx == n || stack.last[1] > heights[idx])
      rect = stack.pop
      start_pos = rect[0]
      best = [best, (idx - start_pos) * rect[1]].max
    end

    stack << [start_pos, heights[idx]] if idx != n
  end

  best
end
